### ISSUE TITLE
npm link fails

### ISSUE THREAD

**Issue by dekz** [serverless/serverless#200]
Following these [instructions](https://github.com/jaws-framework/JAWS/blob/master/docs/create_reusable_module.md).

Trying to create the inital global link (in user home dir) of `awsm-images`:

```
~/t/jaws ❯❯❯ cd awsm-images
~/t/j/awsm-images ❯❯❯ npm link                                                                                                         master
|
-
> awsm-images@0.1.0 postinstall /Users/jacob/tmp/jaws/awsm-images
> jaws postinstall awsm-images npm

{ [JawsError: Must be in a JAWS project]
  name: 'JawsError',
  message: 'Must be in a JAWS project',
  messageId: 12 }

npm ERR! Darwin 14.5.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "link"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! code ELIFECYCLE
npm ERR! awsm-images@0.1.0 postinstall: `jaws postinstall awsm-images npm`
npm ERR! Exit status 12
npm ERR!
npm ERR! Failed at the awsm-images@0.1.0 postinstall script 'jaws postinstall awsm-images npm'.
npm ERR! This is most likely a problem with the awsm-images package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     jaws postinstall awsm-images npm
npm ERR! You can get their info via:
npm ERR!     npm owner ls awsm-images
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/jacob/tmp/jaws/awsm-images/npm-debug.log
```

```
5360 info awsm-images@0.1.0 Failed to exec postinstall script
5361 verbose stack Error: awsm-images@0.1.0 postinstall: `jaws postinstall awsm-images npm`
5361 verbose stack Exit status 12
5361 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/lifecycle.js:213:16)
5361 verbose stack     at EventEmitter.emit (events.js:110:17)
5361 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/spawn.js:24:14)
5361 verbose stack     at ChildProcess.emit (events.js:110:17)
5361 verbose stack     at maybeClose (child_process.js:1015:16)
5361 verbose stack     at Process.ChildProcess._handle.onexit (child_process.js:1087:5)
5362 verbose pkgid awsm-images@0.1.0
5363 verbose cwd /Users/jacob/tmp/jaws/awsm-images
5364 error Darwin 14.5.0
5365 error argv "/usr/local/bin/node" "/usr/local/bin/npm" "link"
5366 error node v0.12.7
5367 error npm  v2.11.3
5368 error code ELIFECYCLE
5369 error awsm-images@0.1.0 postinstall: `jaws postinstall awsm-images npm`
5369 error Exit status 12
5370 error Failed at the awsm-images@0.1.0 postinstall script 'jaws postinstall awsm-images npm'.
```

**Comment by joostfarla** [comment]
There seem to be 2 problems:

* When using `npm link` in the module directory, it seems NPM runs the pre/postinstall scripts. Despite the error, the module is linked anyway. 

* When using `npm link awsm-images` inside your project, the postinstall script is fired, but it runs within the module's directory and not within the project. 

For both problems we need to figure out how we can determine the project's directory. For the first problem we could do nothing if no project directory is found. For the second problem, we need to switch the working directory before executing the tasks. When I output the working directory of the postinstall task, this appears to be the physical path of the module, not the location of the symbolic link. This makes it hard to fix. I will try to investigate further!

For now, until we find a solution, I would recommend to leave out the `postinstall` config in your `package.json` and run the `jaws postinstall awsm-images npm` by hand after linking the module in your project.

**Comment by joostfarla** [comment]
@ac360 @doapp-ryanp this might be a reason to build a tiny wrapper around NPM for installing/updating modules? Otherwise I'm afraid this is quite hard to solve.

**Comment by doapp-ryanp** [comment]
First off thanks for bringing this up. I have an idea. I can write code in the post install hook to check to see if the just installed module is in the Global NPM path and maybe even check the timestamp and version. If that is the case I will suppress the error and potentially not run some of the merging. 

I really want to avoid at all costs not having to write wrappers around package managers as that is not only going to be there prone but yet another thing we'll have to make sure we keep forwards and backwards compatible going forward which is never fun. eyesI missing here? I won't get to this till next week unless someone else wants to take a stab at the proposed algorithm above.

**Comment by doapp-ryanp** [comment]
The mobile web screwed up my input box for some reason. I meant to say does anyone see any angles that I am missing here?

**Comment by joostfarla** [comment]
Well, the second problem might be the hardest to solve. When linking the module from your project, NPM runs the `postinstall` script within the module's directory. From there it seems impossible to determine from within which project the `npm install` command is fired, because the working directory is the physical path of the module and not the linked path within `node_modules`.

Any thought on this? If we can think of a solution, I'm willing to do some implementing in the next few days!

**Comment by ac360** [comment]
@joostfarla Both myself and @doapp-ryanp are out promoting JAWS @ Re:Invent this week.  If you can come up with a fix for this, we would be eternally grateful.  We're a bit out of commission at the moment and it is unfortunate we will be promoting JAWS while this critical feature is broken.

**Comment by joostfarla** [comment]
@ac360 @doapp-ryanp first of all, good luck @ re:Invent!! Let's hope this will create a lot of buzz around the project. I'll try to help wherever I can while you're out!

The linking & postinstalling seems to work well when using a normal `npm install`. This is due to the fact that, in this case, the `install` command is run within the project directory as the current working directory. This is different than the `link` command, which runs the `install` command within the module directory. Unfortunately, I can't figure out a way to determine the originating project directory from there, so that's a real problem.

I see 2 possible solutions for now:

* Remove the `jaws postinstall` command from the `scripts` section in `package.json`. After all, this seems to be meant for running post scripts which reside *inside* the module's repository. This would mean the `jaws postinstall` command has to be run by hand after running `npm install` (which has to be documented). I we want this, we have to fix this ASAP, so people build modules the right way. Optionally, to simplify this in the future, some small wrapper around `npm` (or for other platforms) could be built. 

* In the `postinstall` script, check if it is run from within a project directory (simply check if the parent directory is `node_modules`). If not, do nothing. This means the `postinstall` command is automatically triggered when using `npm install`, but has to be run by hand when using `npm link`.

I would opt for the first option, because this doesn't involve workarounds of any kind. This seems to me like the most flexible and future-proof solution.

Please let me know what you think! I would be happy to submit some PR's to fix this soon.

Small sidenote: Shouldn't the CLI command be called `jaws module postinstall` instead of `jaws postinstall`?

**Comment by joostfarla** [comment]
Just thought this over and I changed my mind on what's best for now. I'll submit a PR for option 2 right away! We can always deprecate this in the future if we come up with something better.

**Comment by sixlettervariables** [comment]
Even when following this updated guidance I cannot seem to get my AWSM module to install in a JAWS project:

```sh
$ npm link awsm-sierraecg
awsm-sierraecg@0.0.4 ../.node/lib/node_modules/awsm-sierraecg
/Users/xyz/sierraecg/node_modules/awsm-sierraecg -> ../.node/lib/node_modules/awsm-sierraecg -> /Users/xyz/.node/lib/node_modules/awsm-sierraecg
$ jaws postinstall awsm-sierraecg npm
JAWS: Copying /Users/xyz/sierraecg/node_modules/awsm-sierraecg/awsm to /Users/xyz/sierraecg/aws_modules/sierraecg  
{ [Error: ENOENT: no such file or directory, lstat '/Users/xyz/sierraecg/aws_modules/awsm-sierraecg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'lstat',
  path: '/Users/xyz/sierraecg/aws_modules/awsm-sierraecg' }
```
It seems to strip the `awsm-` qualifier causing downstream calls to fail.

**Comment by dekz** [comment]
My local install doesn't seem to strip out the `awsm` prefix for the module.

```
!~/t/j/jaws-N1PJWSo1e * ❯❯❯ ls aws_modules                                                                                                                   
awsm-users
```

I recall when I accidentally deleted the `aws_modules` folder, the postinstall process did not recreate it for me and I did have a difficult time with `postinstall`

**Comment by boushley** [comment]
@dekz @sixlettervariables You shouldn't need the `awsm-` prefix in both `package.json` and `awsm.json` anymore. I think this was resolved with https://github.com/jaws-framework/JAWS/pull/224

Let me know if you're still having issues.
